src/launch/linux/ioctl.rs

```diff
@@ -50,6 +54,58 @@ pub const LAUNCH_MEASUREMENT: Ioctl<WriteRead, &Command<LaunchMeasure>> =
 /// the ready state.
 pub const LAUNCH_FINISH: Ioctl<WriteRead, &Command<LaunchFinish>> =
     unsafe { Ioctl::classic(ENC_OP) };
+/// Instruct the kernel to pin these pages so that they aren't remapped
+/// into a different page frame which will cause unpredictable results
+/// due to how SEV's keys depend on encrypted regions being static.
+///
+/// It's possible this may become obsolete as there are proposed kernel
+/// patches to make the kernel pin the pages when they're first touched
+/// automatically.
+pub fn encrypt_region(kvm_fd: &impl AsRawFd, region: &[u8]) -> Result<c_uint> {
+    EncryptedRegion::new(region).ioctl(kvm_fd)
+}
+
+#[repr(C)]
+struct EncryptedRegion<'a> {
+    addr: u64,
+    len: u64,
+    _phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> EncryptedRegion<'a> {
+    fn new(region: &'a [u8]) -> Self {
+        Self {
+            addr: region.as_ptr() as _,
+            len: region.len() as _,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn ioctl(self, fd: &impl AsRawFd) -> Result<c_uint> {
+        // iocuddle (correctly) won't allow the creation of a wrapper for
+        // this ioctl where we supply a struct to the kernel  because it
+        // is declared in the kernel as:
+        //
+        //     _IOR(KVMIO, 0xbb, struct kvm_enc_region)
+        //
+        // instead of as:
+        //
+        //     _IOW(KVMIO, 0xbb, struct kvm_enc_region)
+        //
+        // _IOR means the kernel is writing to a struct for us to read,
+        // but the ioctl is meant to be used as _IOW which means *we* give
+        // the kernel a struct for it to read from.
+        extern "C" {
+            fn ioctl(fd: c_int, request: c_ulong, ...) -> c_int;
+        }
+
+        let r = unsafe { ioctl(fd.as_raw_fd(), REG_REGION, &self as *const _ as c_ulong) };
+
+        let res: c_uint = r.try_into().map_err(|_| Error::last_os_error())?;
+
+        Ok(res)
+    }
```

Comment:
Can't you do this with iocuddle classic?

Reply:
No. The kernel's incorrect definition of `_IOR` would constrain me to iocuddle's `Read` direction which means I can't supply the structure to iocuddle for the kernel to read from my values. When I tried to define the iocuddle direction as `Write` or `WriteRead` it resulted in a different ioctl definition (_IOR vs _IOW)